iroha_network/src/lib.rs

```diff
@@ -268,14 +170,57 @@ impl Request {
         Request::new(uri_path, Vec::new())
     }
 
-    /// getter for url
-    pub fn url(&self) -> &str {
-        &self.uri_path[..]
-    }
+    /// Constructs `Request` from `AsyncStream`
+    /// # Errors
+    /// May fail if parsing fails or reading through stream fails
+    pub async fn from_async_stream(stream: &mut impl AsyncStream) -> Result<Self> {
+        async fn read_header<'a, 'b: 'a>(
+            stream: &'b mut impl AsyncStream,
+            buf: &'a mut [u8],
+        ) -> Result<&'a [u8]> {
+            let len = stream.read(buf).await.wrap_err("Failed to read header")?;
+            Ok(&buf[..len])
+        }
+        fn get_crlf(buf: &[u8]) -> Result<usize> {
+            buf.iter()
+                .position(|&b| b == b'\n')
+                .ok_or_else(|| error!("Request shoud contain CRLF sequences"))
+                .map(|n| n + 1)
+        }
+        fn read_uri(buf: &[u8]) -> Result<(String, &[u8])> {
+            let end_uri = get_crlf(buf)?;
+            let uri =
+                String::from_utf8(buf[..end_uri - 2].to_vec()).wrap_err("Failed to decode uri")?;
+            Ok((uri, &buf[end_uri..]))
+        }
+        fn read_payload_len(buf: &[u8]) -> Result<(usize, &[u8])> {
+            let end_len = get_crlf(buf)?;
+            let len = std::str::from_utf8(&buf[..end_len - 2])
+                .wrap_err("Failed to decode length")?
+                .parse::<usize>()
+                .wrap_err("Failed to parse length")?;
+            Ok((len, &buf[end_len..]))
+        }
```

Comment:
I think they can be simply made into private functions inside of `Request`? Might be easier to read.

Reply:
They are local to this method, no need to actually move it as it won't be reused.

Reply:
At least add empty lines between them for readability :)